### PR TITLE
Use VNVisitReachingVNs in more places

### LINKED ISSUE #104489: JIT: nullcheck is not folded for PHI
Seems like global assertprop bails out on the following snippet
```cs
bool Test(bool cond)
{
    object tmp;
    if (cond)
        tmp = "1111";
    else
        tmp = "2222";
    return tmp != null;
}
```
Current codegen:
```asm
; Assembly listing for method Proga:Test(ubyte):ubyte (FullOpts)
G_M26246_IG01:  ;; offset=0x0000
G_M26246_IG02:  ;; offset=0x0000
       mov      rax, 0x20000209798      ; '2222'
       mov      rdx, 0x200002097B8      ; '1111'
       test     cl, cl
       cmovne   rax, rdx
       test     rax, rax
       setne    al
       movzx    rax, al
G_M26246_IG03:  ;; offset=0x0023
       ret      
```
Expected codegen:
```asm
; Assembly listing for method Proga:Test(ubyte):ubyte (FullOpts)
G_M26246_IG01:  ;; offset=0x0000
G_M26246_IG02:  ;; offset=0x0000
       mov      eax, 1
G_M26246_IG03:  ;; offset=0x0023
       ret      
```

### PR DESCRIPTION
**Pull request by EgorBo**
Closes https://github.com/dotnet/runtime/issues/104489

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=824251&view=ms.vss-build-web.run-extensions-tab)

The main idea to walk Phi nodes in `bool ValueNumStore::IsKnownNonNull(ValueNum vn)` so if all phi nodes are not-null, then the phi itself is not null as well. Same for "does need a write barrier" detection.

The rest of the changes are unrelated, but they were needed to mitigate TP regression

### CODE REVIEW

File: src/coreclr/jit/assertionprop.cpp [hunk 1]
```diff
@@ -2322,38 +2322,6 @@ AssertionInfo Compiler::optAssertionGenJtrue(GenTree* tree)
...
-            break;
-        }
-    }
-
-    // All phi arguments are non-null implies phi rhs is non-null.
-    if (isNonNull)
-    {
-        return optCreateAssertion(tree->AsOp()->gtOp1, nullptr, OAK_NOT_EQUAL);
-    }
-    return NO_ASSERTION_INDEX;
```
Comment (EgorBo) [inline src/coreclr/jit/assertionprop.cpp:2354]: No diff change. Also, it effectively repeats `IsKnownNonNull` now which also walks phis.

File: src/coreclr/jit/assertionprop.cpp [hunk 2]
```diff
@@ -2387,20 +2355,9 @@ void Compiler::optAssertionGen(GenTree* tree)
...
-            }
             break;
 
         case GT_IND:
-            // Dynamic block copy sources could be zero-sized and so should not generate assertions.
-            if (tree->TypeIs(TYP_STRUCT))
-            {
-                break;
-            }
-            FALLTHROUGH;
```
Comment (EgorBo) [inline src/coreclr/jit/assertionprop.cpp:2402]: Dynamic blocks are no more. No diffs change.

File: src/coreclr/jit/assertionprop.cpp [hunk 3]
```diff
@@ -5511,63 +5470,11 @@ void Compiler::optImpliedAssertions(AssertionIndex assertionIndex, ASSERT_TP& ac
...
-            if (curAssertion->IsCopyAssertion())
-            {
-                optImpliedByCopyAssertion(curAssertion, iterAssertion, activeAssertions);
-            }
-            else if (iterAssertion->IsCopyAssertion())
-            {
-                optImpliedByCopyAssertion(iterAssertion, curAssertion, activeAssertions);
-            }
-        }
-    }
```
Comment (EgorBo) [inline src/coreclr/jit/assertionprop.cpp:5566]: No diffs change

**Comment by EgorBo** [comment]
@AndyAyersMS @dotnet/jit-contrib PTAL

File: src/coreclr/jit/valuenum.h [hunk 4]
```diff
@@ -594,6 +595,20 @@ class ValueNumStore
...
+            return argVisitor(vn);
+        }
+        return VNVisitReachingVNsWorker(vn, argVisitor);
+    }
+
+private:
+
+    // Helper function for VNVisitReachingVNs
+    template <typename TArgVisitor>
+    VNVisit VNVisitReachingVNsWorker(ValueNum vn, TArgVisitor argVisitor)
```
Comment (EgorBo) [inline src/coreclr/jit/valuenum.h:611]: Here I split `VNVisitReachingVNs` into two functions. In most cases the input is not a phi so we run `argVisitor` just once. So we don't need to allocate an array (on stack)

File: src/coreclr/jit/assertionprop.cpp [hunk 5]
```diff
@@ -5097,20 +5054,22 @@ bool Compiler::optWriteBarrierAssertionProp_StoreInd(ASSERT_VALARG_TP assertions
...
+    auto vnVisitor = [this](ValueNum vn) -> ValueNumStore::VNVisit {
+        if ((vn == ValueNumStore::VNForNull()) || vnStore->IsVNObjHandle(vn))
+        {
+            // No write barrier is required for null or nongc object handles as values
+            return ValueNumStore::VNVisit::Continue;
+        }
+        return ValueNumStore::VNVisit::Abort;
+    };
+
+    if (vnStore->VNVisitReachingVNs(value->gtVNPair.GetConservative(), vnVisitor) == ValueNumStore::VNVisit::Continue)
```
Comment (EgorBo) [inline src/coreclr/jit/assertionprop.cpp:5066]: Basically, if we have `IND(addr) = value` where `value` is either null or an nongc object - we don't need a write-barrier (taking Phi defs into account now).

Similarly, we can walk phis for `addr` itelf but that didn't show any diffs.

File: src/coreclr/jit/assertionprop.cpp (hunk 3 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/assertionprop.cpp:5566]: Seems a little odd that this has no impact... are we not properly maintaining the vn->assertion maps?

File: src/coreclr/jit/valuenum.cpp [hunk 6]
```diff
@@ -3032,6 +3024,17 @@ bool ValueNumStore::GetPhiDef(ValueNum vn, VNPhiDef* phiDef)
     return false;
 }
 
+bool ValueNumStore::IsPhiDef(ValueNum vn) const
```
Comment (AndyAyersMS) [inline src/coreclr/jit/valuenum.cpp:3027]: Header comment?

File: src/coreclr/jit/assertionprop.cpp (hunk 3 above)
Comment (EgorBo) [reply src/coreclr/jit/assertionprop.cpp:5566]: @AndyAyersMS we do maintain it, but it seems that we never arrive here: https://github.com/dotnet/runtime/blob/81efcade9e543a2c03457b3bc87a2bc76b36057e/src/coreclr/jit/assertionprop.cpp#L5891 (usable is always false, also only `O2K_LCLVAR_COPY` is visitted)

File: src/coreclr/jit/valuenum.cpp (hunk 6 above)
Comment (EgorBo) [reply src/coreclr/jit/valuenum.cpp:3036]: Added

File: src/coreclr/jit/assertionprop.cpp (hunk 3 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/assertionprop.cpp:5566]: Ok. I don't understand this aspect of AP very well, so perhaps nothing is wrong, but it still feels like at some point either something broke and we didn't realize, or something rendered all this logic irrelevant (in which case perhaps we can save TP by removing more code elsewhere). Maybe open an issue to investigate?

**Review by AndyAyersMS — APPROVED** [review]

File: src/coreclr/jit/assertionprop.cpp (hunk 3 above)
Comment (EgorBo) [reply src/coreclr/jit/assertionprop.cpp:5566]: Filed https://github.com/dotnet/runtime/issues/108504

**Pull request merged**